wave_lang/kernel/wave/mlir_converter.py

```diff
@@ -0,0 +1,289 @@
+            AddressSpace.SHARED_MEMORY: "shared",
+            AddressSpace.REGISTER: "register",
+        }[addr]
+    # For symbolic address spaces, return as string (will be handled by water_emitter)
+    return str(addr)
+
+
+def _shape_syms(tt):
+    """Extract symbolic shape from type."""
+    syms = []
+    for s in getattr(tt, "symbolic_shape", []) or []:
+        try:
+            syms.append(str(s))
+        except Exception:
+            syms.append("?")
+    return syms
+
+
+def _serialize_trace(
+    trace, save_to_file: bool = False, filename: str | None = None
+) -> dict:
+    # Local import to avoid requiring package context for CLI mode.
+    from ..ops.wave_ops import get_custom, Placeholder
+    from ..lang.wave_types import Memory, Register
+    from ..lang.kernel_buffer import AddressSpace
+    import torch.fx as fx
+
+    nodes = []
+    walked = list(trace.walk())
+    # Assign stable ids.
+    name_to_id: dict[str, int] = {}
+    for i, node in enumerate(walked):
+        name_to_id[getattr(node, "name", f"n{i}")] = i
+
+    for node in walked:
+        custom = get_custom(node)
+
+        # Call infer_type() on the custom object if it has the method
+        if hasattr(custom, "infer_type"):
+            try:
+                custom.infer_type()
+            except Exception as e:
+                # If infer_type fails, continue without it
+                pass
+
+        entry = {
+            "id": name_to_id.get(getattr(node, "name", ""), -1),
+            "name": getattr(node, "name", ""),
+            "op": getattr(node, "op", ""),
+            "tkw_op_name": getattr(custom, "tkw_op_name", "unknown"),
+            "subgraph": getattr(node, "subgraph_name", ""),
+        }
+        # Capture type info for placeholders.
+        if isinstance(custom, Placeholder):
+            t = getattr(custom, "_type", None)
+            if t is not None:
+                if issubclass(t, Memory):
+                    shape_syms = _shape_syms(t)
+                    addr = _addr_str(
+                        getattr(t, "address_space", AddressSpace.GLOBAL_MEMORY)
+                    )
+                    dt = _dtype_str(getattr(t, "dtype", None))
+                    entry["wave_type"] = {
+                        "kind": "memory",
+                        "shape": shape_syms,
+                        "address_space": addr,
+                        "dtype": dt,
+                    }
+                elif issubclass(t, Register):
+                    shape_syms = _shape_syms(t)
+                    dt = _dtype_str(getattr(t, "dtype", None))
+                    entry["wave_type"] = {
+                        "kind": "register",
+                        "shape": shape_syms,
+                        "address_space": "register",
+                        "dtype": dt,
+                    }
+        # Capture result type for other ops when available on custom.
+        t = getattr(custom, "type", None)
+        if isinstance(t, type):
```

Comment:
Is there a point in this check in absence of `else`? Aren't subclass checks enough?